{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# UNIX Quick Reference Guide\n",
    "\n",
    "## Looking at files and moving them around\n",
    "\n",
    "```\n",
    "pwd # Tell me which directory I'm in\n",
    "ls # What else is in this directory\n",
    "ls .. # What is in the directory above me\n",
    "ls foo/bar/ # What is inside the bar directory which is inside the foo/ directory\n",
    "ls -lah foo/ # Give the the details (-l) of all files and folders (-a) using human\n",
    "             # readable file sizes (-h)\n",
    "cd ../.. # Move up two directories\n",
    "cd ../foo/bar # Move up one directory and down into the foo/bar/ subdirectories\n",
    "cp -r foo/ baz/ # Copy the foo/ directory into the baz/ directory\n",
    "mv baz/foo .. # Move the foo directory into the parent directory\n",
    "rm -r ../foo # remove the directory called foo/ from the parent directory\n",
    "find foo/ -name \"*.gff\" # find all the files with a gff extension in the directory foo/\n",
    "```\n",
    "\n",
    "## Looking in files\n",
    "\n",
    "```\n",
    "less bar.bed # scroll through bar.bed\n",
    "grep chrom bar.bed | less -S # Only look at lines in bar.bed which have 'chrom' and\n",
    "                             # don't wrap lines (-S)\n",
    "head -20 bar.bed # show me the first 20 lines of bar.bed\n",
    "tail -20 bar.bed # show me the last 20 lines\n",
    "cat bar.bed # show me all of the lines (bad for big files)\n",
    "wc -l bar.bed # how many lines are there\n",
    "sort -k 2 -n bar.bed # sort by the second column in numerical order\n",
    "awk '{print $1}' bar.bed | sort | uniq # show the unique entries in the first column\n",
    "```\n",
    "\n",
    "## Grep\n",
    "\n",
    "```\n",
    "grep foo bar.bed # show me the lines in bar.bed with 'foo' in them\n",
    "grep foo baz/* # show me all examples of foo in the files immediately within baz/\n",
    "grep -r foo baz/ # show me all examples of foo in baz/ and every subdirectory within it\n",
    "grep '^foo' bar.bed # show me all of the lines begining with foo\n",
    "grep 'foo$' bar.bed # show me all of the lines ending in foo\n",
    "grep -i '^[acgt]$' bar.bed # show me all of the lines which only have the characters\n",
    "                           # a,c,g and t (ignoring their case)\n",
    "grep -v foo bar.bed # don't show me any files with foo in them\n",
    "```\n",
    "\n",
    "## Awk\n",
    "\n",
    "```\n",
    "awk '{print $1}' bar.bed # just the first column\n",
    "awk '$4 ~ /^foo/' bar.bed # just rows where the 4th column starts with foo\n",
    "awk '$4 == \"foo\" {print $1}' bar.bed # the first column of rows where the 4th column is foo\n",
    "awk -F\"\\t\" '{print $NF}' bar.bed # ignore spaces and print the last column\n",
    "awk -F\"\\t\" '{print $(NF-1)}' bar.bed # print the penultimate column\n",
    "awk '{sum+=$2} END {print sum}' bar.bed # print the sum of the second column\n",
    "awk '/^foo/ {sum+=$2; count+=1} END {print sum/count}' bar.bed # print the average of the\n",
    "                                                               # second value of lines starting\n",
    "                                                               # with foo\n",
    "```\n",
    "\n",
    "## Piping, redirection and more advanced queries\n",
    "\n",
    "```\n",
    "grep -hv '^#' bar/*.gff | awk -F\"\\t\" '{print $1}' | sort -u\n",
    "#  grep => -h: don't print file names\n",
    "#          -v: don't give me matching files\n",
    "#          '^#': get rid of the header rows\n",
    "#          'bar/*.gff': only look in the gff files in bar/\n",
    "#  awk => print the first column\n",
    "#  sort => -u: give me unique values\n",
    "\n",
    "awk 'NR%10 == 0' bar.bed | head -20\n",
    "# awk => NR: is the row number\n",
    "#        NR%10: is the modulo (remander) of dividing my 10\n",
    "#        awk is therefore giving you every 10th line\n",
    "# head => only show the first 20\n",
    "\n",
    "awk '{l=($3-$2+1)}; (l<300 && $2>200000 && $3<250000)' exercises.bed\n",
    "# Gives:\n",
    "# contig-2\t201156\t201359\tgene-67\t24.7\t-\n",
    "# contig-4\t245705\t245932\tgene-163\t24.8\t+\n",
    "# Finds all of the lines with features less than 300 bases long which start\n",
    "# after base 200,000 and end before base 250,000\n",
    "# Note that this appears to have the action before the pattern.  This is\n",
    "# because we need to calculate the length of each feature before we use it\n",
    "# for filtering.  If they were the other way around, you'd get the line\n",
    "# immediatly after the one you want:\n",
    "awk '(l<300 && $2>200000 && $3<250000) {l=($3-$2+1); print $0}' exercises.bed\n",
    "# Gives:\n",
    "# contig-2\t201156\t201359\tgene-67\t24.7\t-\n",
    "# contig-2\t242625\t243449\tgene-68\t46.5\t+\n",
    "```\n",
    "\n",
    "## A script\n",
    "\n",
    "```\n",
    "#!/usr/bin/env bash\n",
    "\n",
    "set -e # stop running the script if there are errors\n",
    "set -u # stop running the script if it uses an unknown variable\n",
    "set -x # print every line before you run it (useful for debugging but annoying)\n",
    "\n",
    "if [ $# -ne 2 ]\n",
    "then\n",
    "  echo \"You must provide two files\"\n",
    "  exit 1 # exit the programme (and number > 0 reports that this is a failure)\n",
    "fi\n",
    "\n",
    "file_one=$1\n",
    "file_two=$2\n",
    "\n",
    "if [ ! -f $file_one ]\n",
    "then\n",
    "  echo \"The first file couldn't be found\"\n",
    "  exit 2\n",
    "fi\n",
    "\n",
    "if [ ! -f $file_two ]\n",
    "then\n",
    "  echo \"The second file couldn't be found\"\n",
    "  exit 2\n",
    "fi\n",
    "\n",
    "# Get the lines which aren't headers,\n",
    "# take the first column and return the unique values\n",
    "number_of_contigs_in_one=$(awk '$1 !~ /^#/ {print $1}' $file_one | sort -u | wc -l)\n",
    "number_of_contigs_in_two=$(awk '/^[^#]/ {print $1}' $file_two | sort -u | wc -l)\n",
    "\n",
    "if [ $number_of_contigs_in_one -gt $number_of_contigs_in_two ]\n",
    "then\n",
    "  echo \"The first file had more unique contigs than the second\"\n",
    "  exit\n",
    "elif [ $number_of_contigs_in_one -lt $number_of_contigs_in_two ]\n",
    "then\n",
    "  echo \"The second file had more unique contigs\"\n",
    "  exit\n",
    "else\n",
    "  echo \"The two files had the same number of contigs\"\n",
    "  exit\n",
    "fi\n",
    "```\n",
    "\n",
    "\n",
    "## Pro tips\n",
    "\n",
    "* Use tab completion - it will save you time!\n",
    "* Always have a quick look at files with `less` or `head` to double check their format\n",
    "* Watch out for data in headers and that you don't accidentally grep some if you don't want them\n",
    "* Watch out for spaces, especially if you're using awk; if in doubt, use `-F\"\\t\"`\n",
    "* Regular expressions are wierd, build them up slowly bit by bit\n",
    "* If you did something smart but can't remember what it was, try typing `history` and it might have a record\n",
    "* `man the_name_of_a_command` often gives you help\n",
    "* Google is normally better at giving examples (prioritise stackoverflow.com results, they're normally good)\n",
    "\n",
    "## Build commands slowly\n",
    "\n",
    "If you wanted me to calculate the sum of all of the scores for genes on contig-1 in a bed file, I'd probably run each of the following commands before moving onto the next:\n",
    "\n",
    "```\n",
    "head -20 bar.bed # check which column is which and if there are any headers\n",
    "head -20 bar.bed | awk '{print $5}' # have a look at the scores\n",
    "awk '{print $1}' bar.bed | sort -u | less # check the contigs don't look wierd\n",
    "awk '{print $4}' bar.bed | sort -u | less # check the genes don't look wierd\n",
    "awk '$4 ~ /gene-/' bar.bed | head -20 # check that I can spot genes\n",
    "awk '($1 == \"contig-1\" && $4 ~ /gene-/)' bar.bed | head -20 # check I can find\n",
    "                                                            # genes on contig-1\n",
    "# check my algorithm works on a subset of the data\n",
    "head -20 bar.bed | awk '($1 == \"contig-1\" && $4 ~ /gene-/) {sum+=$5}; END {print sum}'\n",
    "# apply the algorithm to all of the data\n",
    "awk '($1 == \"contig-1\" && $4 ~ /gene-/) {sum+=$5}; END {print sum}' bar.bed\n",
    "```\n",
    "\n",
    "## Which tool should I use?\n",
    "\n",
    "You should probably use awk if:\n",
    "\n",
    "* your data has columns\n",
    "* you need to do simple maths\n",
    "\n",
    "You should probable use grep if:\n",
    "\n",
    "* you're looking for files which contain some specific text (e.g. `grep -r foo bar/`: look in all the files in `bar/` for any with the word 'foo')\n",
    "\n",
    "You should use find if:\n",
    "\n",
    "* you know something about a file (like it's name or creation date) but not where it is\n",
    "* you want a list of all the files in a subdirectory and its subdirectories etc.\n",
    "\n",
    "You should write a script if:\n",
    "\n",
    "* your code doesn't fit on one line\n",
    "* it's doing something you might want to do again in 3 months\n",
    "* you want someone else to be able to do it without asking loads of questions\n",
    "* you're doing something sensitive (e.g. deleting loads of files)\n",
    "* you're doing something lots of times\n",
    "\n",
    "You should probably use less or head:\n",
    "\n",
    "* always, you should always use less or head to check intermediary steps in your analysis\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
