{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run Roary to generate a core gene alignment\n",
    "\n",
    "At this stage you should have three GFF files generated by Prokka, each in its own directory. Provided your QC looked alright, you are now ready to run Roary to generate the pangenome. \n",
    "\n",
    "We are going to run Roary twice, first with the default settings, and then using MAFFT to generate a core alignment. For both of these runs we will want all the annotation files in the same directory, so lets take a copy of them to our current directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cp annotated_1368*/*.gff ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Roary with default settings\n",
    "Running Roary with the default settings is very straightforward. All you need to do is run `roary *.gff` and it will run on all the GFF files in the current directory. We want to run Roary twice with different settings, so in order to keep track of our output files from each run we will also specify an output directory for Roary to put the files in. Give the following command a go:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "roary -f run_roary_no_alignment *.gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will run for a minute or two.\n",
    "\n",
    "We will have a closer look at the results in the next section, so for now let us just see that there are some output files in the directroy we asked Roary to create:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ls -l run_roary_no_alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Roary with Mafft\n",
    "In order to create a multiFASTA alignment of the core genes, we will now run Roary again, but this time with some more options.\n",
    "\n",
    "|Option|Description|\n",
    "|------|-----------|\n",
    "|-e|Create a multiFASTA alignment of the core genes|\n",
    "|--mafft|Use with -e to use MAFFT instead of PRANK|\n",
    "|-p|Number of threads to use|\n",
    "  \n",
    "For all usage options, you can have a look at the [Roary website](https://sanger-pathogens.github.io/Roary/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "roary -f run_roary_alignment -e --mafft -p 8 *.gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will take a bit longer to run than the previous command, about 5 minutes. Once finished you should have a directory called run_roary_alignment containing the output files, this time including a core_gene_alignment.aln file as well. Just quickly check that this is the case and then we will head over to the [next section](results.ipynb) to have a closer look at the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ls -l run_roary_alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also go [here](qc.ipynb) to revisit the previous section, or go back to the [index page](index.ipynb)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
